verl/workers/roles/critic.py

```diff
@@ -0,0 +1,288 @@
+
+            # pad and slice the inputs if sp > 1
+            if ctx["ulysses_sequence_parallel_size"] > 1:
+                input_ids_rmpad, position_ids_rmpad, pad_size = ulysses_pad_and_slice_inputs(
+                    input_ids_rmpad, position_ids_rmpad, sp_size=self.ulysses_sequence_parallel_size
+                )
+                ctx["pad_size"] = pad_size
+
+            inputs["input_ids"] = input_ids_rmpad
+            inputs["attention_mask"] = None
+            inputs["position_ids"] = position_ids_rmpad
+
+            ctx["indices"] = indices
+            ctx["seqlen"] = seqlen
+            ctx["bs"] = bs
+            return inputs, ctx
+
+        def postprocess_fn_with_rmpad(outputs, ctx):
+            response_length = ctx["response_length"]
+            use_value_head_model = ctx["use_value_head_model"]
+            if use_value_head_model:
+                # For trl.AutoModelForCausalLMWithValueHead
+                values_rmpad = outputs[2].squeeze(0).unsqueeze(-1)
+            else:
+                values_rmpad = outputs.logits
+                values_rmpad = values_rmpad.squeeze(0)  # (total_nnz)
+
+            # gather output if sp > 1
+            if ctx["ulysses_sequence_parallel_size"] > 1:
+                values_rmpad = gather_outpus_and_unpad(
+                    values_rmpad, gather_dim=0, unpad_dim=0, padding_size=ctx["pad_size"]
+                )
+
+            # pad it back
+            values = pad_input(values_rmpad, indices=ctx["indices"], batch=ctx["bs"], seqlen=ctx["seqlen"]).squeeze(-1)
+            values = values[:, -response_length - 1 : -1]
+            return values, ctx
+
+        self.use_remove_padding = self.config.model.get("use_remove_padding", False)
+        if self.use_remove_padding:
+            return preprocess_fn_with_rmpad, postprocess_fn_with_rmpad
+        else:
+            return preprocess_fn_without_rmpad, postprocess_fn_without_rmpad
+
+    @register(dispatch_mode=Dispatch.DP_COMPUTE_PROTO)
+    @DistProfiler.annotate(color="cyan")
+    def compute_values(self, data: DataProto):
+        # Support all hardwares
+        data = data.to(get_device_id())
+        micro_batch_size = self.config.forward_micro_batch_size_per_gpu
+        data.meta_info["micro_batch_size"] = micro_batch_size
+        data.meta_info["max_token_len"] = self.config.forward_max_token_len_per_gpu
+        data.meta_info["use_dynamic_bsz"] = self.config.use_dynamic_bsz
+        preprocess_fn, postprocess_fn = self.get_microbatch_process_fn()
+        ctx = self.engine.get_default_ctx()
+
+        with self.engine.eval_mode():
+            data = self.engine.shard_data(data=data)
+            output, _ = self.engine.infer_batch(
+                data, ctx=ctx, preprocess_fn=preprocess_fn, postprocess_fn=postprocess_fn
+            )
+            output = self.engine.unshard_data(data=output)
+        output = output.to("cpu")
+        return output
+
+    @register(dispatch_mode=Dispatch.DP_COMPUTE_PROTO)
```

Comment:
We will need to decide how we dispatch here.

Reply:
can you elaborate on this?

Reply:
we'll work on that for megatron backend